sound/soc/sof/ipc4-pcm.c

```diff
@@ -37,73 +64,84 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
 
 		/*
-		 * set pipeline state for both FE and BE pipelines for RUNNING state.
-		 * For PAUSE/RESET, set the pipeline state only for the FE pipeline.
+		 * skip triggering pipelines that are already in the requested state. This will
+		 * be true for backend pipelines that are triggered in the BE DAI trigger.
 		 */
-		switch (state) {
-		case SOF_IPC4_PIPE_PAUSED:
-		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
-				continue;
-			break;
-		default:
-			break;
-		}
+		if (pipeline->state != state)
+			data.pipeline_ids[data.count++] = pipeline_info->pipelines[i]->instance_id;
+	}
+
+	/* return if all pipelines are in the requested state already */
+	if (!data.count)
+		return 0;
+
+	/* pause all pipelines */
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, SOF_IPC4_PIPE_PAUSED, &data);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to pause all pipelines\n");
+		return ret;
+	}
 
-		/* find pipeline widget for the pipeline that this widget belongs to */
-		pipeline_widget = swidget->pipe_widget;
-		pipeline = (struct sof_ipc4_pipeline *)pipeline_widget->private;
-
-		if (pipeline->state == state)
-			continue;
-
-		/* first set the pipeline to PAUSED state */
-		if (pipeline->state != SOF_IPC4_PIPE_PAUSED) {
-			ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id,
-							  SOF_IPC4_PIPE_PAUSED);
-			if (ret < 0) {
-				dev_err(sdev->dev, "failed to pause pipeline %d\n",
-					swidget->pipeline_id);
-				return ret;
+	/* update PAUSED state for all pipelines that were just triggered */
+	for (i = 0; i < pipeline_info->count; i++) {
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
+
+		for (j = 0; j < data.count; j++) {
+			if (data.pipeline_ids[j] == pipeline_info->pipelines[i]->instance_id) {
+				pipeline->state = SOF_IPC4_PIPE_PAUSED;
+				break;
 			}
 		}
+	}
 
-		pipeline->state = SOF_IPC4_PIPE_PAUSED;
+	/* return if this is the final state */
+	if (state == SOF_IPC4_PIPE_PAUSED)
+		return ret;
 
-		if (pipeline->state == state)
-			continue;
+	/* else set the final state in the DSP */
+	ret = sof_ipc4_set_multi_pipeline_state(sdev, state, &data);
+	if (ret < 0) {
+		dev_err(sdev->dev, "failed to set final state %d for all pipelines\n", state);
+		return ret;
+	}
 
-		/* then set the final state */
-		ret = sof_ipc4_set_pipeline_state(sdev, pipeline_widget->instance_id, state);
-		if (ret < 0) {
-			dev_err(sdev->dev, "failed to set state %d for pipeline %d\n",
-				state, swidget->pipeline_id);
-			break;
-		}
+	/* update final state for all pipelines that were just triggered */
+	for (i = 0; i < pipeline_info->count; i++) {
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
 
-		pipeline->state = state;
+		for (j = 0; j < data.count; j++) {
+			if (data.pipeline_ids[j] == pipeline_info->pipelines[i]->instance_id) {
```

Comment:
you don't need to check this here since if a pipeline is not in data.pipeline_ids,  pipeline->state should be equal to state in above 
````c
		if (pipeline->state != state)
			data.pipeline_ids[data.count++] = pipeline_info->pipelines[i]->instance_id;
````